doc/developer/design/20230814_unbilled_replicas.md

```diff
@@ -0,0 +1,166 @@
+* Record cluster profile properties in audit events.
+
+## Non-Goals
+
+We do not plan to do the following:
+* Allow customers general access to cluster profiles.
+
+## Overview
+
+We introduce the notion of cluster profiles, which capture groups
+of replica sharing configuration aspects. Profiles can co-exist, and a default
+profile captures the current configuration.
+
+At this time, we only support cluster profiles for internal users, but do not
+give customers the permission to directly interact with cluster profiles. This
+is to give us time to validate the feature, and decide on reserved names.
+
+## Detailed description
+
+We extend clusters to support multiple replica configurations, and name each a profile.
+A cluster profile describes the properties of a set of replicas within a cluster.
+A cluster can have several profiles, and each replica is part of a single profile.
+
+
+```sql
+-- Creating a managed cluster profile without billing
+CREATE CLUSTER PROFILE cluster.support WITH (SIZE '3xlarge', BILLED = FALSE);
+-- Drop the profile, releasing any associated resource.
+DROP CLUSTER PROFILE cluster.support;
+
+-- Alternatively, create an unmanaged profile without billing.
+CREATE CLUSTER PROFILE cluster.support WITH (MANAGED = FALSE, BILLED = FALSE);
+-- Create a replica in the profile, identifying the profile in the name.
+CREATE CLUSTER REPLICA cluster.support.r1 SIZE '3xlarge';
+-- Drop the profile
+DROP CLUSTER PROFILE cluster.support CASCADE;
+```
+
+### Default profile
+
+When a user doesn't specify a profile, they interact with the default profile:
+
+```sql
+CREATE CLUSTER clname SIZE '3xlarge';
+-- Equivalent to:
+CREATE CLUSTER clname;
+CREATE CLUSTER PROFILE clname.default SIZE '3xlarge';
+```
+
+### Resources within profiles
+
+Replicas are part of a profile, and their lifetime is tied to that of the profile:
+
+```sql
+-- Cluster `cl` has a profile `default` with one replica `r1`
+DROP CLUSTER PROFILE cl.default;
+ERROR: Cluster profile cl.default has objects
+HINT: Drop objects first, or specify CASCADE.
+
+DROP CLUSTER PROFILE cl.default CASCADE;
+```
+
+### Managed and unmanaged profiles
+
+We can mix managed and unmanaged profiles:
+
+```sql
+-- Create cluster without profiles
+CREATE CLUSTER cl;
+-- Create `default` profile:
+CREATE CLUSTER PROFILE cl.default SIZE '3xsmall', REPLICATION FACTOR 2;
+-- Create `p1` profile:
+CREATE CLUSTER PROFILE cl.p1 REPLICAS (r1 SIZE 'large');
+```
+
+This is equivalent to creating the default profile in-line:
+
+```sql
+-- Create default profile as part of the cluster:
+CREATE CLUSTER cl SIZE '3xsmall', REPLICATION FACTOR 2;
+-- Create `p1` profile:
+CREATE CLUSTER PROFILE cl.p1 REPLICAS (r1 SIZE 'large');
+```
+
+
+We need to extend the syntax to dropping replicas of unmanaged profiles:
+
+```sql
+-- Drop replica from the default profile:
+DROP CLUSTER REPLICA cl.r1;
+-- Drop replica from specific profile:
+DROP CLUSTER REPLICA cl.p1.r1;
```

Comment:
I'm personally not a huge fan of this syntax. Also having replicas with the same name in the same cluster seems like a recipe for confusion. Have you considered disallowing duplicate names? Then the existing syntax for dropping replicas would be sufficient.

Reply:
Oh, wow, that could be a nice way to make this less disruptive! Instead of changing the structure of names, we could make clusters work like schemas, where they contain multiple different types of objects—in this case, replicas _and_ profiles. The names of all replicas and profiles would need to be unique. So `c.r1` could be a replica or it could be a profile, for example.

Reply:
Interesting! This might complicate the `ALTER CLUSTER clname SET (MANAGED)` implementation for changing an unmanged to a managed cluster, but I'll explore it and see where it leads us.